### PR TITLE
hack: explicitly control enabling the journald logging driver

### LINKED ISSUE #47770: hack/make.sh: control explicitly whether or not to enable the journald logging driver
### Description

Hi all, I am a docker maintainer on Gentoo Linux.
When docker is built from source, `hack/make.sh` always seems to look for libsystemd and link it if it is available.
See this [bug](https://bugs.gentoo.org/914076) for the details.
There should be a build option that controls whether or not this happens.
This issue exists in docker 26.1.0.

### PR DESCRIPTION
**Pull request by williamh**
Without this, the dependency on systemd is said to be "automagic", which can lead to breakage, for example, if a binary package of docker is built on a system that has systemd installed then installed on a system that does not have systemd installed.

for example: https://bugs.gentoo.org/914076

closes https://github.com/moby/moby/issues/47770

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

```markdown changelog

```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by tianon** [comment]
Sorry, I'm actually struggling to understand here.  If you want to build a binary that works for users who may or may not be using systemd (and thus may or may not want the journald logging driver), you need to compile against libsystemd and libsystemd is then a dependency of that binary (and binary package).  The "fix" here is to accurately make `libsystemd` a dependency of the published binary package, not to remove the `journald` logging driver entirely from the builds?

**Comment by tianon** [comment]
> This broke the binary packages for half our users.

... but the fix broke functionality of the binary for the other half :sob:

**Comment by thesamesam** [comment]
> > This broke the binary packages for half our users.
> 
> ... but the fix broke functionality of the binary for the other half 😭

No, it didn't, because we wired up the packaging to pass a flag to enable/disable systemd based on a packaging option.

What we want is a toggle supported upstream to say "yes, I know systemd is installed on this system, please don't use it" to produce a binary which isn't linked against libsystemd, rather than deciding purely based on whether it's installed. The default doesn't matter for us either way, I presume you'd want it on.

**Comment by eli-schwartz** [comment]
@tianon your understanding is fatally flawed.

Gentoo has a robust mechanism for linking feature flags on the distro package to compiled in features on the binary. In order for this to work, the docker build script needs to expose the option.

The current state of affairs is that Gentoo's binary caching layer marks some binaries as being linked to systemd and other binaries as being NOT linked to systemd, but the docker build system is ignoring that and building both against systemd.

End result: people disable the Gentoo feature flag "systemd" but accidentally get a broken binary that was linked against systemd just because it happened to be installed on a buildbot VM.

**Comment by williamh** [comment]
I updated the patch to have this on by default as @thesamesam  suggested above.

**Comment by thaJeztah** [comment]
Looks like the commit is missing a DCO sign-off, causing CI to fail. If you update, could you also remove the `closes #xxx` from the commit message? (it's fine to keep it in the PR description, but my experience is that having them in the commit message sometimes causes a lot of noise on GitHub when forks merge, cherry-pick, or otherwise handle the commit 😓)

**Comment by eli-schwartz** [comment]
Counterpoint: it also makes it harder to tell from the git log, where the related discussion is.

**Comment by williamh** [comment]
I added the DCO signoff and removed the Closes keyword as requested.
@tianon  please take a look.
Thanks for your time.

**Comment by williamh** [comment]
Ping, what is the status of this pr? Thanks much.

File: hack/make.sh
Comment (tianon) [inline hack/make.sh:86]: Instead of the very generic `SYSTEMD` name here, can we use something more descriptive/specific?  I would suggest something like `NO_AUTO_BUILDTAG_JOURNALD` but open to other ideas/suggestions (especially opinions from other maintainers).  Also, `[[` isn't necessary here (`[` is sufficient).

```suggestion
if [ -z "${NO_AUTO_BUILDTAG_JOURNALD:-}" ] && ${PKG_CONFIG} 'libsystemd' 2> /dev/null; then
```

Comment (eli-schwartz) [reply hack/make.sh:86]: > Also, `[[` isn't necessary here (`[` is sufficient).

The flip side is that the script explicitly uses bash, and the `[[` ***keyword*** has no downsides but allows bash to process it much earlier, in time to know that variables are variables -- the single `[` is a builtin program that is also available at `/usr/bin/[` and only knows that it received "argv arguments". This is why the single `[` needs you to quote variables that may be empty or contain spaces, while the `[[` allows omitting the quotes. In fact, even if you omit them by accident, it still works. :)

Comment (tianon) [reply hack/make.sh:86]: Sure, this is more of a consistency request than an argument for one or the other (the rest of `hack/` minus three instances is consistently using bare `[`, which given we're in Bash, also doesn't shell out).

**Comment by williamh** [comment]
@tianon  I used the variable name EXCLUDE_AUTO_BUILDTAG_JOURNALD to make the name consistent with the way other drivers are excluded in docker.

**Review by cpuguy83 — APPROVED** [review]

**Review by vvoland — APPROVED** [review]

**Review by tianon — APPROVED** [review]

**Pull request merged**
